File: product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        bool flag = true;
        int ans=1,ans0=1;
        int count=0;
        for(int i=0;i<nums.size();i++){
            if(nums[i]==0)
            {   
                count++;
                flag=false;
            }
            else
                ans0*=nums[i];
                ans*=nums[i];
        }
        if(flag){
            for(int i=0;i<nums.size();i++){
               int x=nums[i];
                nums[i]=ans/x;
            }
        }
        else if(!flag)
        {
            for(int i=0;i<nums.size();i++){
                if(nums[i]==0)
                    nums[i]=ans0;
                else
                    nums[i]=0;
            }
        }
        if (count>=2)
        {
            for(int i=0;i<nums.size();i++){
                nums[i]=0;
            }
        }
        return nums;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pre = []
        suf = []
        p1 = 1
        p2 = 1
        ans = []
        for i in range(len(nums)):
            p1 *= nums[i]
            p2 *= nums[len(nums)-1-i]
            pre.append(p1)
            suf.append(p2)
        suf = suf[::-1]
        for i in range(len(nums)):
            ans.append((pre[i-1] if i>0 else 1)*(suf[i+1] if i<len(nums)-1 else 1))
        return ans